zebra-network/src/peer_set/initialize.rs

```diff
@@ -72,45 +72,39 @@ where
     let (peerset_tx, peerset_rx) = mpsc::channel::<PeerChange>(100);
     // Create an mpsc channel for peerset demand signaling.
     let (mut demand_tx, demand_rx) = mpsc::channel::<()>(100);
+    let (handle_tx, handle_rx) = tokio::sync::oneshot::channel();
 
     // Connect the rx end to a PeerSet, wrapping new peers in load instruments.
-    let peer_set = Buffer::new(
-        PeerSet::new(
-            PeakEwmaDiscover::new(
-                ServiceStream::new(
-                    // ServiceStream interprets an error as stream termination,
-                    // so discard any errored connections...
-                    peerset_rx.filter(|result| future::ready(result.is_ok())),
-                ),
-                config.ewma_default_rtt,
-                config.ewma_decay_time,
-                NoInstrument,
+    let peer_set = PeerSet::new(
+        PeakEwmaDiscover::new(
+            ServiceStream::new(
+                // ServiceStream interprets an error as stream termination,
+                // so discard any errored connections...
+                peerset_rx.filter(|result| future::ready(result.is_ok())),
             ),
-            demand_tx.clone(),
+            config.ewma_default_rtt,
+            config.ewma_decay_time,
+            NoInstrument,
         ),
-        config.peerset_request_buffer_size,
+        demand_tx.clone(),
+        handle_rx,
     );
 
     // Connect the tx end to the 3 peer sources:
 
     // 1. Initial peers, specified in the config.
-    tokio::spawn(add_initial_peers(
+    let add_guard = tokio::spawn(add_initial_peers(
         config.initial_peers(),
         connector.clone(),
         peerset_tx.clone(),
     ));
 
     // 2. Incoming peer connections, via a listener.
-    tokio::spawn(
-        listen(config.listen_addr, listener, peerset_tx.clone()).map(|result| {
-            if let Err(e) = result {
-                error!(%e);
-            }
-        }),
-    );
+    let listen_guard = tokio::spawn(listen(config.listen_addr, listener, peerset_tx.clone()));
 
     // 3. Outgoing peers we connect to in response to load.
 
+    let peer_set = Buffer::new(peer_set, config.peerset_request_buffer_size);
```

Comment:
could this be moved back up to the place where the `peer_set` is constructed, so it's all in one place? it's easier to miss when it's separated from the rest of the peer set setup.